### PR TITLE
mypy compatibility

### PR DESCRIPTION
**Pull request by phobson**
- [ ] Closes #xxxx
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

This mostly ignore remaining complex typing situations to get mypy passing so that we can enable it and start enforcing going forward. 

Builds off of https://github.com/dask/dask/pull/8706

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:1078]: This is wrong. I think the original author meant to say that you either give a list of sources and a list of targets, or a single source and a single target, but they didn't think about mix and matching. The code however is blindly assuming the two inputs are aligned and that's why mypy is - rightfully! - complaining.
```suggestion
    if isinstance(targets, Array):
        targets = [targets]
```

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:1012]:
```suggestion
    regions: tuple[slice, ...] | Collection[tuple[slice, ...] | None] | None = None,
```
See below comment at line 1090

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:1090]: So regions can be a Collection[None]. This should be reflected in the type declaration; with the type: ignore you're instead completely neutering mypy checks for the use case below.
```suggestion
        regions = [regions]
```

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:1155]:
```suggestion
        load_store_dsk: HighLevelGraph | dict = store_dsk
```
Avoid unnecessary abstraction

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:4128]: You're explicitly stating that this function demands to mutate the input parameters.
This should be treated as a red flag.

One of the big perks of mypy is that it lets you spot parameter poisoning.
Input collections should be labelled as generic read-only whenever it is reasonable to do so.
This however means that you may need to rename internal variables, e.g.

```python
def f(x):
    x = list(x)
    x.append(123)
    return x
```
becomes
```python
def f(x: Collection) -> list:
    out = list(x)
    out.append(123)
    return out
```

Comment (crusaderky) [reply dask/array/core.py:4128]: In this specific case, however, the docstring and the implementation call for an array-like, which numpy does specify; sadly there's no way to label it as immutable.

Add to dask/array/numpy_compat.py:

```python
if _numpy_120:
    from numpy.typing import ArrayLike
else:
    ArrayLike = Any
```

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:4178]: see comment for load_store_chunk

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:4182]: see comment for load_store_chunk

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:4258]:
```suggestion
) -> dict[Hashable, Any]:
```

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:5560]: Return type of ``__init__`` should be omitted
```suggestion
    def __init__(self, array: Array):
```

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:5571]: mypy is telling that something's wrong here.
According to the definition:
``
index: int | Sequence[int] | slice | Sequence[slice]
``
after this block,
- if index was a int, new_index is now a ``tuple[int, ...]`` - good
- if index was a slice, new_index is now a ``tuple[slice, ...]`` - good
- if index was a Sequence but not a tuple, new_index is now a tuple of sequences. Which is actually intentionally in the implementation below.

If you defined new_index explicitly, you would save yourself a wealth of ``#type: ignore`` and benefit from more robust checks:
``new_index: tuple[int | slice, ...] | tuple[Sequence[int | slice]]``

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:5572]: Something's fishy here too. If the purpose was to accept ``np.ndarray | list``, why did the definition go for a much more relaxed Sequence?

File: dask/array/optimization.py
Comment (crusaderky) [inline dask/array/optimization.py:123]: what was wrong here?

File: dask/base.py
Comment (crusaderky) [inline dask/base.py:19]: please import Sequence from collections.abc instead

File: dask/base.py
Comment (crusaderky) [inline dask/base.py:977]: if the return type of normalize_function was wrong, why aren't you changing the type of function_cache too?

File: dask/base.py
Comment (crusaderky) [inline dask/base.py:985]: As a general rule,
- be as generic as possible for parameters
- be as specific as possible for return types

``Sequence[Callable] | Sequence[Any]`` is the same as ``Sequence[Any]``.

I can't see any use case where the output is a Callable.

I see below that one of the possible return types is a ``str | bytes``. While ``str | bytes`` is _technically_ a ``Sequence[Any]``, it's something that should be called out explicitly.
```suggestion
) -> tuple | str | bytes:
```

**Review by crusaderky — CHANGES_REQUESTED** [review]
Partial review - more later

File: dask/array/core.py
Comment (crusaderky) [reply dask/array/core.py:4128]: The return type here is quite painful, as it forces you to check ``not None`` even when you stated ``return_stored: True``.

This is a common issue, with two canonical solutions to this:

1. just return ``Any``. 
```python
def load_store_chunk(
    x: ArrayLike,
    out: ArrayLike,
    index: slice,
    lock: Any,
    return_stored: bool,
    load_stored: bool,
) -> Any:
    [implementation]
```
3. Use overload:
```python
@overload
def load_store_chunk(
    x: ArrayLike,
    out: ArrayLike,
    index: slice,
    lock: Any,
    return_stored: Literal[True],
    load_stored: bool,
) -> ArrayLike:
    ...

@overload
def load_store_chunk(
    x: ArrayLike,
    out: ArrayLike,
    index: slice,
    lock: Any,
    return_stored: Literal[False],
    load_stored: bool,
) -> None:
    ...

def load_store_chunk(x, out, index, lock, return_stored, load_stored):
    [implementation]
```
Neither is perfect and you should decide case by case which is best - to neuter mypy checks in the user code, or to have a very long override definition here. In this case I would opt for the former.

Comment (crusaderky) [reply dask/array/core.py:4128]: p.s. the implementation says that x should never be None

**Comment by mrocklin** [comment]
Before we start requiring mypy in CI I think that we should raise that as a discussion in https://github.com/dask/community (unless there was already a decision on this).

File: dask/array/core.py
Comment (phobson) [reply dask/array/core.py:1090]: Thanks for catching this -- I'm getting hung on this section. Removing the ignore comment and adding `Collection[None]` to the type declariation brings up:

```
```

File: dask/array/optimization.py
Comment (phobson) [reply dask/array/optimization.py:123]: I can't remember. I was having trouble getting it to work originally, but some of the other changes must have benefitted this section. Removing the ignore comments, mypy is satisfied with this section now.

**Comment by crusaderky** [comment]
> Before we start requiring mypy in CI I think that we should raise that as a discussion in https://github.com/dask/community (unless there was already a decision on this).

I believe there was consensus at the maintainers weekly meeting

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:1010]: This is strongly misaligned with the docstring; which is right?

File: dask/array/core.py
Comment (crusaderky) [reply dask/array/core.py:1090]: It works fine for me with
```python
    regions: tuple[slice, ...] | Collection[tuple[slice, ...] | None] | None = None,
```
It won't work with
```python
    regions: tuple[slice, ...] | Collection[tuple[slice, ...]] | Collection[None] | None = None,
```

**Comment by crusaderky** [comment]
I'm a bit swarmed right now with PyConDE and Easter coming up, so I'm afraid I won't be able to continue the review before April 19th

File: dask/blockwise.py
Comment (ian-r-rose) [inline dask/blockwise.py:426]: This can be fixed by removing `name=dep` above, and having an `else` block below where `name = dep`. Then at all times `name` will be a string.

File: dask/blockwise.py
Comment (ian-r-rose) [inline dask/blockwise.py:427]: This can be fixed by annotating above as a `MutableMapping`. But that also should be telling us that we should be making a copy of `numblocks` as a dict internally.

File: dask/dataframe/core.py
Comment (ian-r-rose) [inline dask/dataframe/core.py:8]: With a `from __future__ import annotations`, you should just be able to use the builtin `set`

File: dask/dataframe/tests/test_dataframe.py
Comment (ian-r-rose) [inline dask/dataframe/tests/test_dataframe.py:2358]: Can you add a TODO` to type `from_delayed`? This papercut is annoying enough that I'd like to try to fix it at some point

File: dask/highlevelgraph.py
Comment (ian-r-rose) [inline dask/highlevelgraph.py:7]: I think this can be imported from `collections.abc`

File: dask/layers.py
Comment (ian-r-rose) [inline dask/layers.py:1278]: Hah

File: dask/highlevelgraph.py
Comment (ian-r-rose) [inline dask/highlevelgraph.py:397]: Probably better to just type the input as `dict`. But not worth spending a bunch of time here

File: dask/highlevelgraph.py
Comment (ian-r-rose) [inline dask/highlevelgraph.py:755]: Nice

File: dask/highlevelgraph.py
Comment (ian-r-rose) [inline dask/highlevelgraph.py:408]: This can be fixed by adding some overloads to `keys_in_task` to get the return types right. It's okay to ignore for now, but perhaps add a TODO?

**Review by ian-r-rose — COMMENTED** [review]
Thanks for your work on this @phobson !

File: dask/utils.py
Comment (ian-r-rose) [inline dask/utils.py:1349]: This sure seems like a mypy bug to me

File: dask/array/core.py
Comment (phobson) [reply dask/array/core.py:1016]: This one is interesting. The `None` that I added is inside the `Collection[...]`, meaning (as I see it) a sequence of Nones. When I take this out, mypy complains about this line:

https://github.com/dask/dask/blob/d7a64d723134b9c51e0e38cbe0a94f13349ffddd/dask/array/core.py#L1094-L1095

File: dask/array/core.py
Comment (phobson) [reply dask/array/core.py:4179]: Thanks for this detailed comment and information. I'm about to get on my connecting flight, but I'll try to give this more thought soon.

File: dask/highlevelgraph.py
Comment (phobson) [reply dask/highlevelgraph.py:397]: cool -- going to leave this as-is for now

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:1083]: targets can be
- *any* array-like that supports slice-insertion. Specifically, *not* a dask array!
- a dask delayed
- a collection thereof

All sorts of unit tests should go red with this change, and if they don't it should be worrying.

File: dask/array/core.py
Comment (crusaderky) [reply dask/array/core.py:1016]: This is ugly. In order to make the implementation work, you made the input spec more lax. This could be avoided by just using a different variable name.

File: dask/array/core.py
Comment (crusaderky) [inline dask/array/core.py:1087]: mypy was complaining that you missed *all* use cases from the signature.

File: dask/dataframe/core.py
Comment (crusaderky) [inline dask/dataframe/core.py:3358]:
```suggestion
    _accessors: ClassVar[set[str]] = set()
```
There is no benefit in using an abstract type here.

File: dask/dataframe/core.py
Comment (crusaderky) [inline dask/dataframe/core.py:4030]:
```suggestion
    _accessors: ClassVar[set[str]] = set()
```

File: dask/dataframe/utils.py
Comment (crusaderky) [inline dask/dataframe/utils.py:14]: It was right before:

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.api.types.is_categorical_dtype.html?highlight=is_categorical_dtype

Public documentation wins over type annotations, generally.
If mypy complains, then it's a bug in the pandas typing annotations and it should be reported.

File: dask/graph_manipulation.py
Comment (crusaderky) [inline dask/graph_manipulation.py:9]: please use collections.abc.Set instead

File: dask/highlevelgraph.py
Comment (crusaderky) [inline dask/highlevelgraph.py:397]: 1. If the input is actually a read-only mapping, this will return... another read-only mapping that you can't write to on line 409.
2. there is no reason to spend extra CPU for a deep copy.
```suggestion
        dependencies = ensure_dict(known_key_dependencies, copy=True)
```

File: dask/highlevelgraph.py
Comment (crusaderky) [inline dask/highlevelgraph.py:409]: I guess mypy was saying "you can't update a read only mapping". mypy is absolutely right with that deepcopy.
```suggestion
        dependencies.update(
```

File: dask/layers.py
Comment (crusaderky) [inline dask/layers.py:8]: please use collections.abc.Callable instead

**Review by crusaderky — CHANGES_REQUESTED** [review]

File: dask/highlevelgraph.py
Comment (phobson) [reply dask/highlevelgraph.py:1060]: Without this, we get the following error:
```
dask/highlevelgraph.py:1059: error: Argument 2 to "__dask_distributed_pack__" of "Layer" has incompatible type "Dict[Hashable, AbstractSet[Any]]"; expected "Mapping[Hashable, Set[Any]]"
```

Setting the type hint for `self.key_dependencies` to:
```
    key_dependencies: dict[Hashable, Set] | Layer
```
in the class definition and `__init__` signature only brings up more errors. I'll come back to this, but right now, I'm not able to figure out what's needed to make this work.

**Comment by phobson** [comment]
Took a stab at addressing a few more comments (the easily ones, of course).

Currently getting the following failures locally

### unit tests

```
_______________________________________________ test_store_nocompute_regions _______________________________________________
dask/array/tests/test_array_core.py:2020: in test_store_nocompute_regions
    d1 = da.store(x, y, regions=(0,), compute=False)
dask/array/core.py:1092: in store
    raise ValueError(
E   ValueError: Different number of sources [1] and targets [2]
_____________________________________________________ test_store_locks _____________________________________________________
dask/array/tests/test_array_core.py:2091: in test_store_locks
    assert store(a, at, scheduler="threads", num_workers=10) is None
dask/array/core.py:1091: in store
    if len(sources) != len(targets):
E   TypeError: object of type 'NonthreadSafeStore' has no len()
```


### pre-commit (mypy)

```
dask/array/core.py:1098: error: List item 0 has incompatible type "Union[Tuple[slice, ...], Tuple[Any, ...], None]"; expected "Tuple[slice, ...]"
dask/array/core.py:1168: error: Incompatible types in assignment (expression has type "Dict[Tuple[Any, ...], Any]", variable has type "Union[HighLevelGraph, Dict[Hashable, Any]]")
```

File: .pre-commit-config.yaml
Comment (phobson) [reply .pre-commit-config.yaml]: replaced `typing_extension.Literal` with `typing.Literal`

File: dask/highlevelgraph.py
Comment (phobson) [reply dask/highlevelgraph.py]: Thanks @crusaderky looks like that worked 👍🏻

File: dask/dataframe/core.py
Comment (phobson) [reply dask/dataframe/core.py:4090]: You're right about that. However, without the `# type: ignore`, I get the following:

```
dask/dataframe/core.py:4091: error: Read-only property cannot override read-write property
```

Which is decidedly different that the comment I have. I should have (I think) referred to this issue:
https://github.com/python/mypy/issues/4125

I propose I update the link in the comment and keep this type-ignored for now.

Comment (crusaderky) [reply dask/dataframe/core.py]: yes please

File: dask/dataframe/utils.py
Comment (phobson) [reply dask/dataframe/utils.py]: Mypy isn't complaining any more. Reverted this.

File: dask/base.py
Comment (phobson) [reply dask/base.py]: I've unignored this, fixed (simplified) the type annotations, and at the root of the issue I updated the annotation for the `function_cache`. Should be good to go with this one.

File: dask/base.py
Comment (phobson) [reply dask/base.py]: Thanks this set me on the right path (I believe).

File: dask/graph_manipulation.py
Comment (crusaderky) [reply dask/graph_manipulation.py]: nope. That's fed from HighLevelGraph.dependencies, which (very much deliberately) states that everything must be read-only.

**Comment by crusaderky** [comment]
Final code review at https://github.com/phobson/dask/pull/2

**Comment by phobson** [comment]
Failed Mac build(s) appear to be related to GHA issues.

**Review by crusaderky — APPROVED** [review]

**Comment by crusaderky** [comment]
Thank you!

**Comment by ian-r-rose** [comment]
Woo! Thanks @phobson and @crusaderky!

**Comment by phobson** [comment]
Thanks @crusaderky and @ian-r-rose  for the patience and guidance!

**Pull request merged**
